### ISSUE TITLE
get_unwrap changes mutability

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#9909]
### Summary

.

### Lint Name

get_unwrap

### Reproducer

I tried this code:

```rust
fn linidx(row: usize, col: usize) -> usize {
    row * 1 + col * 3
}

fn main() {
    let mut mat = [1.0f32, 5.0, 9.0, 2.0, 6.0, 10.0, 3.0, 7.0, 11.0, 4.0, 8.0, 12.0];

    for i in 0..2 {
        for j in i+1..3 {
            if mat[linidx(j, 3)] > mat[linidx(i, 3)] {
                    for k in 0..4 {
                            let (x, rest) = mat.split_at_mut(linidx(i, k) + 1);
                            let a = x.last_mut().unwrap();
                            let b = rest.get_mut(linidx(j, k) - linidx(i, k) - 1).unwrap();
                            ::std::mem::swap(a, b);
                    }
            }
        }
    }
    assert_eq!([9.0, 5.0, 1.0, 10.0, 6.0, 2.0, 11.0, 7.0, 3.0, 12.0, 8.0, 4.0], mat);
}
```

I saw this happen:
` cargo clippy --fix -- -Aclippy::all -Wclippy::get_unwrap`
```
    Checking clpy v0.1.0 (/tmp/clpy)
warning: failed to automatically apply fixes suggested by rustc to crate `clpy`

after fixes were automatically applied the compiler reported errors within these files:

  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0308]: mismatched types
   --> src/main.rs:16:49
    |
16  | ...                   ::std::mem::swap(a, b);
    |                       ----------------    ^
    |                       |                   |
    |                       |                   expected `&mut f32`, found `f32`
    |                       |                   help: consider mutably borrowing here: `&mut b`
    |                       arguments to this function are incorrect
    |
note: function defined here
   --> /home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/mem/mod.rs:719:14
    |
719 | pub const fn swap<T>(x: &mut T, y: &mut T) {
    |              ^^^^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
Original diagnostics will follow.

warning: called `.get_mut().unwrap()` on a slice. Using `[]` is more clear and more concise
  --> src/main.rs:15:37
   |
15 | ...                   let b = rest.get_mut(linidx(j, k) - linidx(i, k) - 1).unwrap();
   |                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try this: `rest[linidx(j, k) - linidx(i, k) - 1]`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#get_unwrap
   = note: requested on the command line with `-W clippy::get-unwrap`
```

### Version

```text
rustc 1.67.0-nightly (c5d82ed7a 2022-11-19)
binary: rustc
commit-hash: c5d82ed7a4ad94a538bb87e5016e7d5ce0bd434b
commit-date: 2022-11-19
host: x86_64-unknown-linux-gnu
release: 1.67.0-nightly
LLVM version: 15.0.4
```
